Added checks into polynomials.py

jennifercw opened the pull request:
Addressed TODOs requiring checks that values are integers or natural numbers.

jennifercw committed:
Added checks into polynomials.py

Addressed TODOs requiring checks that values are integers or natural
numbers.

smichr commented on sympy/functions/special/polynomials.py:
```diff
@@ -52,7 +52,7 @@ def _eval_conjugate(self):
 
 
 class jacobi(OrthogonalPolynomial):
-    r"""
```
I think you should leave the `r` here: it marks this as a "raw" docstring with backslashes needing to not be interpreted as the beginning of an escape sequence.

smichr commented on sympy/functions/special/polynomials.py:
```diff
@@ -162,7 +162,9 @@ def eval(cls, n, a, b, x):
                 if n.is_positive:
-                    # TODO: Make sure a+b+2*n \notin Z
+                    # Make sure a+b+2*n \notin Z
+                    if (a + b + 2*n) % 1 == 0:
```
or, using assumptions (maybe the preferred way): `if (a + b + 2*n).is_integer:`

smichr commented on sympy/functions/special/polynomials.py:
```diff
@@ -198,7 +200,9 @@ def fdiff(self, argindex=4):
         from sympy import Sum
-        # TODO: Make sure n \in N
+        # Make sure n \in N
+        if n % 1 != 0 or n < 0:
```
-> `if not n.is_integer or n.is_negative:`

smichr commented on sympy/functions/special/polynomials.py:
```diff
@@ -1111,7 +1115,9 @@ def fdiff(self, argindex=2):
     def _eval_rewrite_as_polynomial(self, n, x):
         from sympy import Sum
-        # TODO: Should make sure n is in N_0
+        # Make sure n is in N_0
```
following syntax above "n \in N_0" and then update test to use assumptions and the same for the change below (which I have not added a comment to).

jennifercw committed:
Changed checks to use assumptions

Checks have been changed in response to comments on previous commit

jennifercw committed:
Changed comment

Changed form of comment (missed this whilst making previous changes)

asmeurer commented:
I think it would be better to check `n.is_integer is False` rather than `not n.is_integer`. The latter will also trigger for `n.is_integer == None`, meaning it is not known or not specified. The way it is now, just using `n = Symbol('n')` would not work. You'd have to force it to be an integer each time with `n = Symbol('n', integer=True)`. Worse, if the assumptions system can't deduce that something is an integer, it would fail, even if it really is. Using `n.is_integer is False` avoids this because it only fails when `n` is known to not be an integer.

jennifercw committed:
Changed not n.is_integer to n.is_integer is False

Changed not n.is_integer to n.is_integer is False to avoid this
evaluating to true when n.is_integer == None

smichr commented:
> better to check

agreed

smichr commented:
And now if you add 4 tests that trigger these exceptions, this will be done. Something like `raises(ValueError, lambda: expr_that_will_fail)` in the appropriate `test_foo` file.

jennifercw committed:
Added tests for checks

Added tests to check that ValueErrors are raised

jennifercw commented:
I have added tests for the cases where we must check n is a natural number (I am new to this method of testing so apologies if I've made any mistakes there).
I apologise, but I was having some difficulty working out what kind of expression would trigger the a+b+2*n \notin Z ValueError?

smichr commented:
> would trigger the a+b+2*n

How about `(Dummy(positive=True), 1, 2, oo)`?

jennifercw commented:
Okay, I will put that in. What is the proper way to set something to infinity?

jennifercw committed:
Added test for Jacobi

smichr commented on sympy/functions/special/tests/test_spec_polynomials.py:
```diff
@@ -52,6 +52,9 @@ def test_jacobi():
                                     /((a + b + 2*n + 1)*factorial(n)*gamma(a + b + n + 1))))
 
+    raises(ValueError, lambda: jacobi(-2.1, a, b, x))
+    raises(ValueError, lambda: jacobi(Dummy(positive = True), 1, 2, oo))
```
nit: for keywords, PEP8 suggests removing the space from around the `=`.

jennifercw committed:
Removed spaces around '='

jennifercw committed:
Changed dummy parameters

Test for Jacobi was failing previously. I believe this should fix it

jennifercw commented:
I can't work out why the Jacobi test appears to be failing? It is all working fine for me

toolforger commented:
It's possible that your local mirror of the ```master``` branch is outdated and you have the ```Dummy``` symbol but Travis (which is working off the current branch) does not have it anymore in that module.

The ```slow``` and ```Theano``` runs stalled, that could be due to download problems or some tests simply taking longer on Travis than at home.

jennifercw commented:
Would it be appropriate for me to change the test with 'Dummy' in it to `jacobi(Symbol('y', integer=True, positive=True), 1, 2, oo)`?

smichr commented:
> Would it be appropriate

yes - perhaps name it `n` since that is the parameter it represents. Alternatively you can just add `Dummy` to the import line at the top of the test file (after Symbol, for example). The test fails because Dummy was not imported.

jennifercw committed:
Added import for Dummy

jennifercw commented:
Is there anything else that needs to be done to this or can someone merge it?

jennifercw commented:
Is there anything I need to do to ensure that this is merged before the deadline?
